docs/language/informal/covariant-overrides.md

```diff
@@ -0,0 +1,742 @@
+
+*   The **desired type** is the type that the user wants to use inside the body
+    of the method. This is the real type that the overridden method requires in
+    order to function. Here, it's `RadioButton`.
+
+    The desired type is a subtype of the original type. Going from the original
+    type to the desired type is a downcast, which means the cast needs to be
+    checked and may fail at runtime.
+
+    Even though the desired type is ensconced in the body of the method in the
+    manual pattern, it really is part of the method's signature. If you are
+    calling `addChild()` *on an object that you statically know is a
+    RadioGroup*, you want the errors as if its declared type was the tighter
+    desired type, not the original one. This is something the manual pattern
+    can't express.
+
+So we need to understand the original type, the desired type, and when
+this feature comes into play. To enable it on a method parameter, you
+mark it with the contextual keyword `covariant`:
+
+```dart
+class Widget {
+  void addChild(covariant Widget widget) {...}
+}
+```
+
+Doing so says "A subclass may override this parameter with a tighter desired
+type". A subclass can then override it like so:
+
+```dart
+class RadioGroup extends Widget {
+  void addChild(RadioButton button) {
+    ...
+  }
+}
+```
+
+No special marker is needed in the overriding definition. The presence of
+`covariant` in the superclass is enough. The parameter type in the base class
+method becomes the *original type* of the overridden parameter. The parameter
+type in the derived method is the *desired type*.
+
+This approach fits well when a developer provides a library or framework where
+some parameter types were designed for getting tightened. For instance, the
+`Widget` hierarchy was designed like that.
+
+In cases where the supertype authors did not foresee this need, it is still
+possible to tighten a parameter type by putting the `covariant` modifier on
+the *overriding* parameter declaration.
+
+In general, a tightened type for a parameter `p` in a method `m` is allowed
+when at least one of the overridden declarations of `m` has a `covariant`
+modifier on the declaration corresponding to `p`.
+
+The `covariant` modifier can also be used on mutable fields. Doing so
+corresponds to marking the parameter in the implicitly generated setter
+for that field as `covariant`:
+
+```dart
+class Widget {
+  covariant Widget child;
+}
+```
+
+This is syntactic sugar for:
+
+```dart
+class Widget {
+  Widget _child;
+  Widget get child => _child;
+  set child(covariant Widget value) { _child = value; }
+}
+```
+
+### Grammar changes
+
+We update the following grammar rules:
+
+```
+normalFormalParameter:
+  "covariant"? functionSignature |
+  fieldFormalParameter |
+  simpleFormalParameter
+  ;
+
+simpleFormalParameter:
+  metadata "covariant"? finalConstVarOrType? identifier
```

Comment:
I've introduced a new `functionFormalParameter` to work as the variant of `functionSignature` which is intended to declare a formal parameter. That one needs to admit `covariant` on the parameter declaration as a whole, but other usages of `functionSignature` (e.g., for declaring a top-level function) should not. We can't just add `covariant` in front of a `functionSignature` in the rule for `normalFormalParameter`, because it would then put the metadata in an inconsistent location.

Next, I have made `covariant` a **built-in identifier**, which I believe is accepted at this point. So `covariant` cannot be a type, and hence we cannot have the function header `void foo(covariant covariant covariant);`.

This means that we do not need to disambiguate unconditionally when seeing `covariant` in the parser, which means that we _do_ allow omitting the type (`void foo(covariant x);`) and using `covariant` as the name of the parameter (`void foo(covariant);` and `void foo(bool covariant);`).

As a consequence, we interpret `int time(covariant f())` as having a covariant parameter named `f` whose type is `dynamic Function()`.